module/zfs/spa.c

```diff
@@ -2316,6 +2320,176 @@ vdev_count_verify_zaps(vdev_t *vd)
+	    &orig_txg);
+	(void) nvlist_lookup_uint64(config, ZPOOL_CONFIG_TIMESTAMP,
+	    &orig_timestamp);
+	dprintf("mmp spa_activity_check_reqd orig_txg %llu "
+	    "orig_timestamp %llu\n", (u_longlong_t)orig_txg,
+	    (u_longlong_t)orig_timestamp);
+
+	(void) nvlist_lookup_uint64(config, ZPOOL_CONFIG_POOL_STATE,
+	    &config_pool_state);
+	(void) nvlist_lookup_uint64(config, ZPOOL_CONFIG_HOSTID, &hostid);
+	myhostid = spa_get_hostid();
+
+	dprintf("mmp spa_load_impl pool_state %llu hostid %llu myhostid %llu\n",
+	    config_pool_state,
+	    (u_longlong_t)hostid,
+	    (u_longlong_t)myhostid);
+
+	/*
+	 * Special values to indicate test should be skipped, for zdb
+	 */
+	if (orig_txg == MMP_SKIP_ACTIVITY_TEST ||
+	    spa->spa_skip_activity_chk == B_TRUE)
+		return B_FALSE;
+
+	/*
+	 * Since the hostid is likely to be different before the pivot than
+	 * it is after the pivot, and since it is hard to see how two nodes
+	 * could safely boot from the same pool anyway, we do not do MMP
+	 * for root pools.
+	 */
+	if (spa_is_root(spa))
+		return B_FALSE;
+
+	/*
+	 * Skip the test if the current host, and the host who last imported the
+	 * pool, both have (or had) MMP disabled.  If the pool is successfully
+	 * imported, we will then respect the local setting for mmp writes.
+	 */
+	if (ub->ub_mmp_magic == MMP_MAGIC && ub->ub_mmp_delay == 0 &&
+	    zfs_mmp_interval == 0)
+		return (B_FALSE);
+	/*
+	 * If the orig_* values are nonzero, they are the results of an earlier
+	 * tryimport.  If they match the uberblock we just found, then the pool
+	 * has not changed and we return false so we do not test a second time.
+	 */
+	if (orig_txg && orig_timestamp && orig_txg == ub->ub_txg &&
+	    orig_timestamp == ub->ub_timestamp)
+		return (B_FALSE);
+
+	if (hostid == myhostid)
+		return (B_FALSE);
+
+	if (config_pool_state != POOL_STATE_ACTIVE)
+		return (B_FALSE);
+
+	return (B_TRUE);
+}
+
+static int
+spa_activity_check(uberblock_t *ub, spa_t *spa)
+{
+	kcondvar_t mycv;
+	kmutex_t mymtx;
+	hrtime_t duration_factor = 4 * zfs_mmp_fail_intervals;
+	hrtime_t activity_test_duration = duration_factor * zfs_mmp_interval;
+	uint64_t orig_txg = 0;
+	uint64_t orig_timestamp = 0;
+	vdev_t *rvd = spa->spa_root_vdev;
+	int error = 0;
+	nvlist_t *label;
+
+	cv_init(&mycv, NULL, CV_DEFAULT, NULL);
+	mutex_init(&mymtx, NULL, MUTEX_DEFAULT, NULL);
+	mutex_enter(&mymtx);
+
+	orig_txg = ub->ub_txg;
+	orig_timestamp = ub->ub_timestamp;
+	if (ub->ub_mmp_magic == MMP_MAGIC && ub->ub_mmp_delay) {
+		/*
+		 * ub_mmp_delay * vdev_count_leaves() == zfs_mmp_interval
+		 * on the node that last had pool imported.
+		 */
+		activity_test_duration = duration_factor * ub->ub_mmp_delay *
+		    vdev_count_leaves(spa);
+	}
+
+	/* Randomize in case of simultaneous imports */
+	activity_test_duration += spa_get_random(3) *
+	    activity_test_duration / 10;
+
+	activity_test_duration += gethrtime();
+	for (int i = 0; gethrtime() < activity_test_duration ; i++) {
+		if (i > 0) {
```

Comment:
I don't think it can.  The rest of the loop is executed for every value of i.  This code (the cv_timedwait IIRC) is executed for every i > 0, so that we only wait in between calls to vdev_load_uberblock.

Maybe I'm missing something - if so, let me know.